src/backend/distributed/planner/fast_path_router_planner.c

```diff
@@ -0,0 +1,409 @@
+	{
+		return false;
+	}
+
+	/* if that's a reference table, we don't need to check anything further */
+	distributionKey = PartitionColumn(distributedTableId, 1);
+	if (!distributionKey)
+	{
+		return true;
+	}
+
+	/* convert list of expressions into expression tree for further processing */
+	quals = joinTree->quals;
+	if (quals != NULL && IsA(quals, List))
+	{
+		quals = (Node *) make_ands_explicit((List *) quals);
+	}
+
+	/*
+	 * Distribution column must be used in a simple equality match check and it must be
+	 * place at top level conjustion operator. In simple words, we should have
+	 *	    WHERE dist_key = VALUE [AND  ....];
+	 *
+	 *	We're also not allowing any other appearances of the distribution key in the quals.
+	 *
+	 *	Overall the logic is might sound fuzzy since it involves two individual checks:
+	 *	    (a) Check for top level AND operator with one side being "dist_key = const"
+	 *	    (b) Only allow single appearance of "dist_key" in the quals
+	 *
+	 *	This is to simplify both of the individual checks and omit various edge cases
+	 *	that might arise with multiple distribution keys in the quals.
+	 */
+	if (ColumnMatchExpressionAtTopLevelConjunction(quals, distributionKey) &&
+		!MultipleColumnsKeysExist(quals, distributionKey))
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * MultipleDistributionKeysExist returns true if the given input
+ * appears more than once in the quals.
+ */
+static bool
+MultipleColumnsKeysExist(Node *quals, Var *distributionKey)
+{
+	ListCell *varClauseCell = NULL;
+	List *varClauseList = NIL;
+	int partitionColumnReferenceCount = 0;
+
+	/* make sure partition column is used only once in the quals */
+	varClauseList = pull_var_clause_default(quals);
+	foreach(varClauseCell, varClauseList)
+	{
+		Var *column = (Var *) lfirst(varClauseCell);
+		if (equal(column, distributionKey))
+		{
+			partitionColumnReferenceCount++;
+
+			if (partitionColumnReferenceCount > 1)
+			{
+				return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * ColumnMatchExpressionAtTopLevelConjunction returns true if the query contains an exact
+ * match (equal) expression on the provided column. The function returns true only
+ * if the match expression has an AND relation with the rest of the expression tree.
+ */
+static bool
+ColumnMatchExpressionAtTopLevelConjunction(Node *node, Var *column)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, OpExpr))
+	{
+		OpExpr *opExpr = (OpExpr *) node;
+		bool distKeyInSimpleOpExpression =
+			DistKeyInSimpleOpExpression((Expr *) opExpr, column);
+
+		if (!distKeyInSimpleOpExpression)
+		{
+			return false;
+		}
+
+		return OperatorImplementsEquality(opExpr->opno);
+	}
+	else if (IsA(node, BoolExpr))
+	{
+		BoolExpr *boolExpr = (BoolExpr *) node;
+		List *argumentList = boolExpr->args;
+		ListCell *argumentCell = NULL;
+
+		if (boolExpr->boolop != AND_EXPR)
+		{
+			return false;
```

Comment:
maybe small comment on happens here, e.g. We do not descend into boolean expressions other than AND. If the column filter appears in an OR clause, we do not consider it even if it is logically the same as a single value comparison (e.g. `<column> = <Const> OR false`).